feat(#1029): Add repository checklist (#1030)
medic/cht-docs: content/en/contribute/code/repository-checklist.md

---
title: "Repository Checklist"

> linkTitle: "Repository Checklist"

Reviewer: Maybe "Repository Creation Checklist" is better?

Reviewer 2: This is a great point, thanks @garethbowen! We should consider a rewrite.

However, I think I have a more nascent concern for both product and repositories: We should always strive to help a real customer/solve a real problem as early as possible.  For example, if [CHT Monitoring](https://github.com/medic/cht-monitoring), which I would consider a product, needed to meet all the "at least do the following" items, we wouldn't be live now and it may be weeks or months until we met all the requirements on that list.  I don't think the investment in checking all those boxes would be worth if it because A) the lost time on helping actual customers and B) we may have a failed product in 3 months and the month we spent doing the checklist was a waste.

The list is actually great though!  I think we can do the rewrite Gareth suggested and greatly reduce the "must haves" and introduce a  "nice to have" list maybe?  The must have list might be:

* The main branch is locked via [branch protection rules](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/managing-a-branch-protection-rule).
* Merges are done through PRs.
* PRs reference related issues.
* Commit formats follow the [guidelines](http://localhost:1313/contribute/code/workflow/#commits).
* Secrets are not part of the commit history or made public.
* For public repositories the main branch should contain a `license` and `readme.md`
*  `main` branch is always shippable.
*  A minimum number of reviewers (usually 2) for a PR merge is enforced by policy

weight: 5
description: >
  Checklist to consider when creating repositories under Medic's GitHub account
---

## Repository Checklist
When creating a new repository under [Medic's GitHub organization](https://github.com/medic), the contributor(s) should at least do the following:

### Source Control
- [ ] The `main` branch is locked via [branch protection rules](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/managing-a-branch-protection-rule).
- [ ] Merges are done through PRs.
- [ ] Issue templates exist.
- [ ] PR template exists.
- [ ] PRs reference related issues.
- [ ] Commit formats follow the [guidelines]({{< ref "workflow/#commits" >}}). 
- [ ] Secrets are not part of the commit history or made public.

> - [ ] For public repositories the `main` branch should contain the following files:

Reviewer 2: A bit silly to have a license for a private repo, but it's really nice to have both, always:

```suggestion
- [ ] For repositories the `main` branch should contain the following files:
```

    - `LICENSE` ([example](https://github.com/medic/cht-core/blob/master/LICENSE))
    - `README.md`

### Testing

> - [ ] Unit tests cover the majority of the code (>90% if possible).

Reviewer: We've discussed internally and haven't agreed on coverage targets. This is probably a separate conversation. To keep this change uncontroversial, let's drop it.

```suggestion
- [ ] Unit tests cover the majority of the code.
```

- [ ] If applicable, integration tests run to test the solution e2e.

### CI/CD
- [ ] Repo runs GitHub Actions CI with automated build and test on each PR.

> - [ ] `Main` branch is always shippable.

Reviewer: ```suggestion
- [ ] `main` branch is always shippable.
```

> - [ ] If applicable, repo uses CD to manage deployments to a development environment before PRs are merged.

Reviewer: Not sure about this one. So far the plans are for CD for commits to `main` which pushes to the production market. Why use CD before merge?

Author: CD can be set up to deploy the changes to a development environment for testing before merging the branch to `main` and deploying the changes to production. I agree that we are not there yet, so I will remove this item from the list.

### Code Reviews
- [ ] The PR template contains a code review checklist.

> - [ ] A minimum number of reviewers (usually 2) for a PR merge is enforced by policy.

Reviewer: ```suggestion
- [ ] A reviewer for a PR merge is enforced by policy.
```

- [ ] Linters, unit tests and successful builds for PR merges are set up.

### Observability
- [ ] Application faults and errors are logged.
- [ ] Logging configuration can be modified without code changes (eg: verbose mode).

## Medic GitHub repository FAQ

### Q: Who can create a repository?

> A: TODO

Reviewer 2: Seconded! Low barrier to entry solves problems now.  Let's rescind access when we have a problem later.

> A: TODO

Author: I assume "Anyone at Medic", right? Can people outside of Medic create repositories under Medic organization?

### Q: When to create a repository?
A: TODO

### Q: Is it OK to create a CHT/Medic-related work repository under a personal GitHub account?

> A: TODO

Reviewer 2: Agreed!  Also, if they follow the "Always have an OSS license", even less of an issue

---
title: "CHT Product Repository Checklist"
linkTitle: "CHT Product Repository Checklist"
weight: 5
description: >
  Checklist to consider when creating CHT Product repositories under Medic's GitHub organization account
---

## Repository Creation Checklist
When creating a new CHT Product repository under [Medic's GitHub organization](https://github.com/medic), the contributor(s) should use the `cht-template` (TODO add link) repository containing the following configurations:

### Source Control
- [ ] The `main` branch is locked via [branch protection rules](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/managing-a-branch-protection-rule).
- [ ] Merges are done through PRs.
- [ ] Issue templates exist.
- [ ] PR template exists.
- [ ] PRs reference related issues.
- [ ] Commit formats follow the [guidelines]({{< ref "workflow/#commits" >}}). 
- [ ] Secrets are not part of the commit history or made public.

> - [ ] The `main` branch contains the following files:

Reviewer: This applies to the repo in general not just the main branch

```suggestion
- [ ] The following files:
```

>     - `LICENSE` ([example](https://github.com/medic/cht-core/blob/master/LICENSE))

Reviewer: ```suggestion
    - `LICENSE` specifying AGPL-3.0 ([example](https://github.com/medic/cht-core/blob/master/LICENSE))
```

## Items to consider when developing the CHT Product
To ensure quality, the CHT Products should also follow the guidelines below:

### CI/CD
- [ ] Repository runs GitHub Actions CI with automated build and test on each PR.

### Testing
- [ ] Unit tests and successful builds for PR merges are set up.
- [ ] Unit tests cover the majority of the code.
- [ ] If applicable, integration tests run to test the solution e2e.

### Observability
- [ ] Application faults and errors are logged.
- [ ] Logging configuration can be modified without code changes (eg: verbose mode).

## Medic GitHub repository FAQ

### Q: Who can create a repository?
A: Anyone under Medic GitHub organization.

### Q: Is it OK to create a CHT/Medic-related work repository under a personal GitHub account?
A: If what you are working on is temporary and just for you then it is fine to create a repository under your personal account (it is the equivalent of having a script on your local machine), as long as it contains an Open-Source Software License. However, default to the Medic account so the other team members can collaborate on it.

### Q: When to make a repo public vs private?

> A: In general, repositories should be public unless there is very good reason to make it private (e.g. the repository contains partner details that cannot be disclosed to public). Always keep in mind that it is much easier to start public than change to public later.

Reviewer: I think we should be more opinionated about open source.

```suggestion
A: Repositories should be public unless there is very good reason to make it private (e.g. the repository contains partner details that cannot be disclosed to public). Always keep in mind that it is much easier to start public than change to public later.
```
